linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -606,13 +613,27 @@ final class Emitter(config: Emitter.Config) {
 
   // Caching
 
-  private final class ModuleCache {
+  private final class ModuleCache extends knowledgeGuardian.KnowledgeAccessor {
     private[this] var _cacheUsed: Boolean = false
 
     private[this] var _importsCache: WithGlobals[List[js.Tree]] = WithGlobals.nil
     private[this] var _lastExternalDependencies: Set[String] = Set.empty
     private[this] var _lastInternalDependencies: Set[ModuleID] = Set.empty
 
+    private[this] var _initializersCache: WithGlobals[List[js.Tree]] = WithGlobals.nil
+    private[this] var _lastInitializers: List[ModuleInitializer.Initializer] = Nil
+
+    override def invalidate(): Unit = {
+      super.invalidate()
+
+      /* The imports cache do not use any global knowledge, so they need not
+       * be invalidated.
+       */
```

Comment:
What is the trade-off here that we need to do this (rather than separating the cache).

Have you considered nesting an `ImportsCache` and `InitializerCache` inside the `ModuleCache` just like we have it for the elements of a class in `ClassCache`?

Reply:
Yeah, this was probably misguided. I added the invalidation of the `importsCache`. It's probably non-consequential, because module initializers and top-level exports should be invalidated rarely.